### PR TITLE
[New Rule] Instance Metadata Service (IMDS) API Requests - Linux

### PR DESCRIPTION
**Pull request by terrancedejesus**
# Pull Request

*Issue link(s)*: https://github.com/elastic/ia-trade-team/issues/272

## Summary - What I changed
Adding a new detection rule for unusual instance metadata service (IMDS) API requests. The IMDS has a global API for cloud service providers that will allow compute instance information retrieval, as well as can facilitate federated authentication and credential access. Adversaries, with compromised credentials or instance, may leverage a custom binary or native tools to send requests to IMDS' API for instance information or event temporary credential requests.

#### Personal stack
Screenshot 2024-08-26 at 11 28 15 AM

#### Telemetry 

Last 4 months there is ~36 alerts that match this. A majority of these are FP for a specific custom shellscript by a client that should be an exception rather than global exclusion in the query.

## Checklist

- [x] Added a label for the type of pr: `bug`, `enhancement`, `schema`, `Rule: New`, `Rule: Deprecation`, `Rule: Tuning`, `Hunt: New`, or `Hunt: Tuning` so guidelines can be generated
- [ ] Added the `meta:rapid-merge` label if planning to merge within 24 hours
- [x] Secret and sensitive material has been managed correctly
- [ ] Automated testing was updated or added to match the most common scenarios
- [ ] Documentation and comments were added for features that require explanation

## Contributor checklist

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md)?

### CODE REVIEW

**Comment by protectionsmachine** [comment]
## Rule: New - Guidelines

These guidelines serve as a reminder set of considerations when proposing a new rule.

### Documentation and Context

- [ ] Detailed description of the rule.
- [ ] List any new fields required in ECS/data sources.
- [ ] Link related issues or PRs.
- [ ] Include references.

### Rule Metadata Checks

- [ ] `creation_date` matches the date of creation PR initially merged.
- [ ] `min_stack_version` should support the widest stack versions.
- [ ] `name` and `description` should be descriptive and not include typos.
- [ ] `query` should be inclusive, not overly exclusive, considering performance for diverse environments. Non ecs fields should be added to `non-ecs-schema.json` if not available in an integration.
- [ ] `min_stack_comments` and `min_stack_version` should be included if the rule is only compatible starting from a specific stack version.
- [ ] `index` pattern should be neither too specific nor too vague, ensuring it accurately matches the relevant data stream (e.g., use logs-endpoint.process-* for process data).
- [ ] `integration` should align with the `index`. If the integration is newly introduced, ensure the manifest, schemas, and `new_rule.yaml` template are updated.
- [ ] `setup` should include the necessary steps to configure the integration.
- [ ] `note` should include any additional information (e.g. Triage and analysis investigation guides, timeline templates).
- [ ] `tags` should be relevant to the threat and align/added to the `EXPECTED_RULE_TAGS` in the definitions.py file.
- [ ] `threat`, `techniques`, and `subtechniques` should map to ATT&CK always if possible.

#### New BBR Rules
- [ ] `building_block_type` should be included if the rule is a building block and the rule should be located in the `rules_building_block` folder.
- [ ] `bypass_bbr_timing` should be included if adding custom lookback timing to the rule.

### Testing and Validation

- [ ] Provide evidence of testing and detecting the expected threat.
- [ ] Check for existence of coverage to prevent duplication.

File: rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            process.name:(
+                curl or wget or python* or perl* or php* or ruby* or lua* or telnet or pwsh or
+                openssl or nc or ncat or netcat or awk or gawk or mawk or nawk or socat or node
+            ) or
+            process.executable: (
+                ./* or /tmp/* or /var/tmp/* or /var/www/* or /dev/shm/* or /etc/init.d/* or
+                /etc/rc*.d/* or /etc/cron* or /etc/update-motd.d/* or /boot/* or /srv/* or
+                /run/* or /etc/rc.local
+            )
+        ) and destination.ip: (169.254.169.254 or "fd00:ec2::254")
```
Comment (Aegrah) [inline rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml:49]: `destination.ip` can't be used with the `process` category. So as we discussed on Slack, you could either choose to be category-less and make fields such as `process.working_directory` and `process.parent.executable` optional, or use a `sequence` with a process --> network event, which allows "best of both worlds".

File: rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+references = ["https://hackingthe.cloud/aws/general-knowledge/intro_metadata_service/"]
+risk_score = 47
+rule_id = "ecc0cd54-608e-11ef-ab6d-f661ea17fbce"
+severity = "medium"
+tags = [
+    "Domain: Endpoint",
+    "OS: Linux",
+    "Use Case: Threat Detection",
+    "Tactic: Credential Access",
+    "Tactic: Discovery",
```
Comment (Aegrah) [inline rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml:29]: There's discovery here, but I don't see it in the MITRE listing

File: rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml (hunk 1 above)
Comment (Aegrah) [reply rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml:49]: One other approach could be to just monitor for network events only, and white list processes + executables. Only risk this approach would have is that it is difficult to tune, given that Linux network events do not have parent fields populated.

File: rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml (hunk 2 above)
Comment (terrancedejesus) [reply rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml:30]: Added

File: rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                "/etc/rc*.d/*", "/etc/cron*", "/etc/update-motd.d/*", "/boot/*", "/srv/*",
+                "/run/*", "/etc/rc.local")
+or process.command_line: "*169.254.169.254*"
+and not process.working_directory: (
+        "/opt/rapid7*",
+        "/opt/nessus*",
+        "/snap/amazon-ssm-agent*",
+        "/srv/snp/docker/overlay2*",
+        "/var/log/amazon/ssm*")] by process.entity_id
+[network where host.os.type == "linux" and event.action == "connection_attempted" and destination.ip == "169.254.169.254"] by process.parent.entity_id
```
Comment (Aegrah) [inline rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml:49]:
```suggestion
sequence by host.id with maxspan=5s
  [process where host.os.type == "linux" and event.type == "start" and event.action == "exec" and process.name : (
    "curl", "wget", "python*", "perl*", "php*", "ruby*", "lua*", "telnet", "pwsh",
    "openssl", "nc", "ncat", "netcat", "awk", "gawk", "mawk", "nawk", "socat", "node"
    ) or process.executable : (
      "./*", "/tmp/*", "/var/tmp/*", "/var/www/*", "/dev/shm/*", "/etc/init.d/*", "/etc/rc*.d/*",
      "/etc/cron*", "/etc/update-motd.d/*", "/boot/*", "/srv/*", "/run/*", "/etc/rc.local"
    ) or 
    process.command_line: "*169.254.169.254*" and
    not process.working_directory: (
      "/opt/rapid7*",
      "/opt/nessus*",
      "/snap/amazon-ssm-agent*",
      "/srv/snp/docker/overlay2*",
      "/var/log/amazon/ssm*"
    )] by process.entity_id
[network where host.os.type == "linux" and event.action == "connection_attempted" and destination.ip == "169.254.169.254"] by process.parent.entity_id
```

**Review by Aegrah — APPROVED** [review]
Added event.type and event.action to the sequence, as the screenshot you provided detects a UID change event followed by a network event, rather than a process execution event. You might want to go back and check whether the query still works as intended with this change.

Other than that, just some formatting. LGTM!

File: rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    ) or 
+    process.command_line: "*169.254.169.254*" and
+    not process.working_directory: (
+      "/opt/rapid7*",
+      "/opt/nessus*",
+      "/snap/amazon-ssm-agent*",
+      "/srv/snp/docker/overlay2*",
+      "/var/log/amazon/ssm*"
+    )] by process.entity_id
+[network where host.os.type == "linux" and event.action == "connection_attempted" and destination.ip == "169.254.169.254"] by process.parent.entity_id
```
Comment (Samirbous) [inline rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml:52]:
```suggestion
sequence by host.id with maxspan=5s
  [process where host.os.type == "linux" and event.type == "start" and event.action == "exec" and 
  (
   process.name : (
    "curl", "wget", "python*", "perl*", "php*", "ruby*", "lua*", "telnet", "pwsh",
    "openssl", "nc", "ncat", "netcat", "awk", "gawk", "mawk", "nawk", "socat", "node"
    ) or 
    
    process.executable : ("./*", "/tmp/*", "/var/tmp/*", "/var/www/*", "/dev/shm/*", "/etc/init.d/*", "/etc/rc*.d/*",
      "/etc/cron*", "/etc/update-motd.d/*", "/boot/*", "/srv/*", "/run/*", "/etc/rc.local") 
    ) and 
    not process.working_directory: (
      "/opt/rapid7*",
      "/opt/nessus*",
      "/snap/amazon-ssm-agent*",
      "/srv/snp/docker/overlay2*",
      "/var/log/amazon/ssm*"
    )] by process.entity_id
[network where host.os.type == "linux" and event.action == "connection_attempted"] by process.parent.entity_id
```

File: rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    "OS: Linux",
+    "Use Case: Threat Detection",
+    "Tactic: Credential Access",
+    "Tactic: Discovery",
+    "Data Source: Elastic Defend",
+]
+type = "eql"
+
+query = '''
+sequence by host.id with maxspan=5s
```
Comment (Samirbous) [inline rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml:36]: where in the query we can see its related to `(IMDS) API Request` ?

Comment (terrancedejesus) [reply rules/linux/credential_access_unusual_instance_metadata_service_api_request.toml]: Must have been removed when committing suggestion. The IMDS relation is either identified in the subsequent network connection event directly to `169.254.169.254` or being used in the command-line.

**Comment by terrancedejesus** [comment]
> Added event.type and event.action to the sequence, as the screenshot you provided detects a UID change event followed by a network event, rather than a process execution event. You might want to go back and check whether the query still works as intended with this change.

joining on `process.entity_id` and `process.parent.entity_id` for the sequences does not match the expected events, hence the image you pointed out. Rather, they entity IDs for both parent and child process are the same so I added `process.parent.entity_id` to the initial `sequence by ...` instead.

Screenshot 2024-08-28 at 12 22 00 PM

**Review by Samirbous — APPROVED** [review]

**Pull request merged**